test/optimizer/optimizer_rule_test.cpp

```diff
@@ -62,6 +65,199 @@ TEST_F(OptimizerRuleTests, SimpleRuleApplyTest) {
+
+  // Setup rule
+  InnerJoinAssociativity rule;
+
+  EXPECT_TRUE(rule.Check(parent_join, root_context));
+  std::vector<std::shared_ptr<OperatorExpression>> outputs;
+  rule.Transform(parent_join, outputs, root_context);
+  EXPECT_EQ(1, outputs.size());
+
+  auto output_join = outputs[0];
+
+  EXPECT_EQ(left_leaf, output_join->Children()[0]);
+  EXPECT_EQ(middle_leaf, output_join->Children()[1]->Children()[0]);
+  EXPECT_EQ(right_leaf, output_join->Children()[1]->Children()[1]);
+
+  auto parent_join_op = output_join->Op().As<LogicalInnerJoin>();
+  auto child_join_op = output_join->Children()[1]->Op().As<LogicalInnerJoin>();
+  EXPECT_EQ(2, parent_join_op->join_predicates.size());
+  EXPECT_EQ(0, child_join_op->join_predicates.size());
+}
+
+TEST_F(OptimizerRuleTests, SimpleAssociativeRuleTest2) {
+  // Start Join Structure: (left JOIN middle) JOIN right
+  // End Join Structure: left JOIN (middle JOIN right)
+  // Query: SELECT * from test1, test2, test3 WHERE test1.a = test3.a AND
+  // test2.a = test3.a; Test ensures that predicate "test2.a = test3.a" is
+  // redistributed to child join
+
+  // Setup Memo
+  Optimizer optimizer;
+  auto &memo = optimizer.metadata_.memo;
+
+  auto left_get = std::make_shared<OperatorExpression>(
+      LogicalGet::make(0, {}, nullptr, "test1"));
+  auto middle_get = std::make_shared<OperatorExpression>(
+      LogicalGet::make(1, {}, nullptr, "test2"));
+  auto right_get = std::make_shared<OperatorExpression>(
+      LogicalGet::make(2, {}, nullptr, "test3"));
+
+  // Create Groups for Get Operators
+  auto left_get_group = memo.InsertExpression(
+      optimizer.metadata_.MakeGroupExpression(left_get), true);
+  auto middle_get_group = memo.InsertExpression(
+      optimizer.metadata_.MakeGroupExpression(middle_get), true);
+  auto right_get_group = memo.InsertExpression(
+      optimizer.metadata_.MakeGroupExpression(right_get), true);
+
+  auto left_leaf = std::make_shared<OperatorExpression>(
+      LeafOperator::make(left_get_group->GetGroupID()));
+  auto middle_leaf = std::make_shared<OperatorExpression>(
+      LeafOperator::make(middle_get_group->GetGroupID()));
+  auto right_leaf = std::make_shared<OperatorExpression>(
+      LeafOperator::make(right_get_group->GetGroupID()));
+
+  // Make Child Join
+  auto child_join =
+      std::make_shared<OperatorExpression>(LogicalInnerJoin::make());
+  child_join->PushChild(left_leaf);
+  child_join->PushChild(middle_leaf);
+  optimizer.metadata_.memo.InsertExpression(
+      optimizer.metadata_.MakeGroupExpression(child_join), true);
+
+  // Make Parent join
+  std::vector<AnnotatedExpression> parent_join_predicates;
+  auto dummy_expr = std::shared_ptr<expression::AbstractExpression>{
+      new expression::OperatorExpression(ExpressionType::COMPARE_EQUAL,
+                                         type::TypeId::INTEGER)};
+  std::unordered_set<std::string> pred1_tables({"test1", "test3"});
+  AnnotatedExpression pred1 = {dummy_expr, pred1_tables};
```

Comment:
I guess I forgot the implementation details here, but there's no column information in the AnnotatedExpression? Where is column information (which column to compare with) stored then?

Reply:
I am unsure, I realize the tests should also account for this. @chenboy do you have an idea?

Reply:
The annotated expression should store the predicate, which contains the column information, and the table alias set. I think it should be fine to not construct the expression in this test because join order enumeration will only look at the table of the used columns, not the expression. For now It's hard for a unit test to actually build & bind an expression. We may want to build some testing infrastructure to make life easier.

Reply:
I see. That's fine for now.